ASoC: SOF: fix return values for mandatory ops

ranj063 opened the pull request:
Fix return value for mandatory ops.

Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

ranj063 commented:
@plbossart this leaves only the PCM ops but I'm not sure whether they are mandotory or not. Could you please comment?

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -229,15 +235,17 @@ static inline int snd_sof_dsp_get_reply(struct snd_sof_dev *sdev,
 
-	return 0;
+	dev_err(sdev->dev, "error: DSP is ipc ready not defined\n");
+	return -ENOTSUPP;
```
is this applied on the right branch, we have both is_ready and is_ipc_ready.

ranj063 commented:
@plbossart fixed now. Could you please cross-check if the list of mandatory and optional ops is correct?

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -67,6 +67,9 @@ struct snd_sof_pdata;
  */
 struct snd_sof_dsp_ops {
+
+	/* The following ops are mandatory */
```
it's not very easy to figure out where the list stops. maybe for each op so that is can be searched whether it's optional or mandatory?

keyonjie commented on sound/soc/sof/sof-priv.h:
```diff
@@ -67,6 +67,9 @@ struct snd_sof_pdata;
+	/* The following ops are mandatory */
+
 	/* probe and remove */
 	int (*remove)(struct snd_sof_dev *sof_dev);
```
I agree we should make .remove mandatory, but the fact is that we don't have it on skl- platforms yet, better to fix them together at one shot(e.g. define an empty one at least).

plbossart replied:
why should .remove be mandatory? if all your driver uses devm_ stuff and the hardware is reset correctly in the .probe, why would we care?

ranj063 replied:
@plbossart going by what remove() does for SKL+, we also disable the DSP in remove(). Given that we are not really unloading the topology from the DSP like we do in the case of load, shouldnt we also be turning off the DSP here?

ranj063 replied:
@plbossart i believe for module load/unload to work in SOF, we should make remove() mandatory so that the DSP can be turned OFF here. Please let me know if you disagree and I'll make changes accordingly.

keyonjie commented on sound/soc/sof/sof-priv.h:
```diff
@@ -151,11 +142,25 @@ struct snd_sof_dsp_ops {
-	int (*load_firmware)(struct snd_sof_dev *sof_dev);
-	int (*load_module)(struct snd_sof_dev *sof_dev,
-			   struct snd_sof_mod_hdr *hdr);
-	int (*fw_ready)(struct snd_sof_dev *sdev, u32 msg_id);
```
IMO, stall, core_power_up/down should be optional, and write, read, write64, read64, block_read, block_write are same for all our platforms so let's remove them from struct snd_sof_dsp_ops.
Others looks fine.

plbossart replied:
it's not because we use the same read/write/write64/read64 that we should remove the abstraction in ops. @dbaluta's last PR shows there could be different ones for other platforms.

ranj063 commented:
> 1. we don't have .remove yet for skl- platforms(please fix that).
> 2. stall, core_power_up/down should be optional.

@keyonjie With SMP enabled in the driver, they are no long optional. 
> 3. write, read, write64, read64, block_read, block_write are same for all our platforms so let's remove them from struct snd_sof_dsp_ops.
> 
> Others looks fine.

keyonjie commented:
> > 1. we don't have .remove yet for skl- platforms(please fix that).
> > 2. stall, core_power_up/down should be optional.
> 
> @keyonjie With SMP enabled in the driver, they are no long optional.

for SKL- platforms, we don't have SMP there, I meant those callbacks are not mandatory for snd_sof_dsp_ops (e.g. no need for byt), but they are mandatory for apl/cnl?

> 
> > 1. write, read, write64, read64, block_read, block_write are same for all our platforms so let's remove them from struct snd_sof_dsp_ops.
> > 
> > Others looks fine.

ranj063 commented:
> > > 1. we don't have .remove yet for skl- platforms(please fix that).
> > > 2. stall, core_power_up/down should be optional.
> > 
> > 
> > @keyonjie With SMP enabled in the driver, they are no long optional.
> 
> for SKL- platforms, we don't have SMP there, I meant those callbacks are not mandatory for snd_sof_dsp_ops (e.g. no need for byt), but they are mandatory for apl/cnl?
> 
> > > 1. write, read, write64, read64, block_read, block_write are same for all our platforms so let's remove them from struct snd_sof_dsp_ops.
> > > 
> > > Others looks fine.

@keyonjie ok Makes sense

ranj063 commented:
> 1. we don't have .remove yet for skl- platforms(please fix that).
> 2. stall, core_power_up/down should be optional.
> 3. write, read, write64, read64, block_read, block_write are same for all our platforms so let's remove them from struct snd_sof_dsp_ops.

@keyonjie hsw-spi.c has a different set of ops!
> 
> Others looks fine.

ranj063 commented:
@keyonjie @plbossart PR updated now.

plbossart commented:
@ranj063 can you rebase, there is conflict. Also please mark each function as mandatory/optional, you are mixing the two styles (per function and per block) which will make searches more complicated.
Thanks!

ranj063 commented:
@plbossart this one has been updated too. There are 2 lines over 80 chars which checkpatch complains about but I think there's no simple way to address it while keeping the style consistent for marking the ops are madatory/optional.

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -330,7 +345,8 @@ snd_sof_pcm_platform_pointer(struct snd_sof_dev *sdev,
 
-	return 0;
+	dev_err(sdev->dev, "error: %s not defined\n", __func__);
+	return -ENOTSUPP;
```
Are PCM operations actually required? Take the example of Sue Creek, you may have all the processing done on the device itself with no PCM being passed back to the host?

ranj063 replied:
@plbossart ok i will make these optional

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -67,100 +67,110 @@ struct snd_sof_pdata;
-	u64 (*read64)(struct snd_sof_dev *sof_dev, void __iomem *addr);
+			u64 value); /* mandatory */
+	u64 (*read64)(struct snd_sof_dev *sof_dev,
+		      void __iomem *addr); /* mandatory */
```
this is inconsistent with the fact that those functions return 0 if not supported?

ranj063 commented:
@plbossart i've updated the PR based on your comments.

wenqingfu commented:
ping @plbossart

plbossart commented:
> ping @plbossart <https://github.com/plbossart>
>
Ranjani and I are working on PM stuff, this can wait.
>

ranj063 commented:
@plbossart I've updated this PR now to address your comment about the is_ready() op. There will be conflicts between this one and my other PR where I have changed the name for this op. I will update them depending on which one gets merged first.

ranj063 commented:
@plbossart please dont merge this. with the return value of undefined values change to < 0, I need to ensure the check is done correctly.

ranj063 commented:
@plbossart this one has been updated now.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -270,8 +271,15 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
+	if (is_dsp_ipc_ready < 0) {
+		spin_unlock_irq(&sdev->ipc_lock);
+		return -EBUSY;
+	}
```
sorry, what does this do?
The is_ready op is marked a mandatory 
int (*is_ready)(struct snd_sof_dev *sof_dev); /* mandatory */

The comment seems wrong?

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -67,100 +67,110 @@ struct snd_sof_pdata;
+	/* FW loading */
+	int (*load_firmware)(struct snd_sof_dev *sof_dev); /* mandatory */
+	int (*load_module)(struct snd_sof_dev *sof_dev,
+			   struct snd_sof_mod_hdr *hdr); /* mandatory */
```
meh, this is not defined for HDAudio.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -67,100 +67,110 @@ struct snd_sof_pdata;
-	int (*reset)(struct snd_sof_dev *sof_dev);
+	int (*run)(struct snd_sof_dev *sof_dev); /* mandatory */
+	int (*stall)(struct snd_sof_dev *sof_dev); /* optional */
+	int (*reset)(struct snd_sof_dev *sof_dev); /* mandatory */
```
meh. not defined for HDaudio.

ranj063 replied:
@plbossart all 3 comments addressed and PR updated now.

plbossart commented on sound/soc/sof/ops.h:
```diff
@@ -152,15 +154,23 @@ static inline void snd_sof_dsp_dbg_dump(struct snd_sof_dev *sdev, u32 flags)
+		return;
+	}
+
+	dev_err(sdev->dev, "error: %s not defined\n", __func__);
```
Ah sorry, I just thought about this one. Since this will not interrupt the flow, we may want to add a _once or _ratelimited qualifier to avoid polluting dmesg. Same remark for all cases where we return void.

ranj063 replied:
@plbossart done
